File: 1Lab/Stack.cpp
#include "Stack.h"
#include <iostream>

using namespace std;

void SCREATE(Stack& s) {
    s.top = nullptr;
    s.size = 0;
    cout << "Стек создан." << endl;
}

bool isEmpty(const Stack& s) {
    return s.top == nullptr;
}

void SPUSH(Stack& s, const string& value) {
    SNode* newNode = new SNode(value, s.top);
    s.top = newNode;
    s.size++;
}

string SPOP(Stack& s) {
    if (isEmpty(s)) {
        cerr << "Ошибка: стек пуст! Невозможно выполнить pop." << endl;
        return "";
    }
    SNode* temp = s.top;
    string popped_value = temp -> data;
    
    s.top = s.top -> next;

    delete temp;
    s.size--;
    return popped_value;
}

string SPEEK(const Stack& s) {
    if (isEmpty(s)) {
        cerr << "Ошибка: стек пуст! Невозможно прочитать элемент." << endl;
        return "";
    }
    return s.top -> data;
}

void SPRINT(const Stack& s) {
    if (isEmpty(s)) {
        cout << "Стек пуст." << endl;
        return;
    }

    cout << "Вершина -> ";
    SNode* current = s.top;
    while (current != nullptr) {
        cout << "\"" << current -> data << "\" ";
        current = current -> next;
    }
    cout << "-> Дно" << endl;
}

void SCLEAR(Stack& s) {
    while (!isEmpty(s)) {
        SPOP(s);
    }
    cout << "Стек очищен." << endl;
}

File: 1LabGo/stack.go
package main

import "fmt"

type StackNode struct {
	data string
	next *StackNode
}

type Stack struct {
	top  *StackNode
	size int
}

func (s *Stack) SPush(data string) {
	newNode := &StackNode{data: data, next: s.top}
	s.top = newNode
	s.size++
}

func (s *Stack) SPop() (string, error) {
	if s.top == nil {
		return "", fmt.Errorf("ошибка: стек пуст")
	}
	data := s.top.data
	s.top = s.top.next
	s.size--
	return data, nil
}

func (s *Stack) SPeek() (string, error) {
	if s.top == nil {
		return "", fmt.Errorf("ошибка: стек пуст")
	}
	return s.top.data, nil
}

func (s *Stack) SPrint() {
	if s.top == nil {
		fmt.Println("Стек пуст.")
		return
	}
	fmt.Print("Вершина -> ")
	current := s.top
	for current != nil {
		fmt.Printf(`"%s" `, current.data)
		current = current.next
	}
	fmt.Println("-> Дно")
}